File: Java/Hard/Accepted/max-sum-of-rectangle-no-larger-than-k_3.java
class Solution {
    public int maxSumSubmatrix(int[][] matrix, int k) {
        
        int rows = matrix.length;
        int cols = matrix[0].length;
        int max = Integer.MIN_VALUE;
        for (int left = 0; left < cols; left++) {
            int[] rowSum = new int[rows];
            for (int right = left; right < cols; right++) {
                for (int i = 0; i < rows; i++) {
                    rowSum[i] += matrix[i][right];
                }
                max = Math.max(max, dpmax(rowSum, k));
                if (max == k) {
                    return k;
                }
            }
        }
        return max;
    }

    private int dpmax(int[] rowSum, int k) {
        int max = Integer.MIN_VALUE;
        int sum = 0;
        TreeSet<Integer> set = new TreeSet<>();
        set.add(0);
        for (int i = 0; i < rowSum.length; i++) {
            sum += rowSum[i];
            Integer ceiling = set.ceiling(sum - k);
            if (ceiling != null) {
                max = Math.max(max, sum - ceiling);
            }
            set.add(sum);
        }
        return max;
    }
}

File: Rust/Hard/Accepted/max-sum-of-rectangle-no-larger-than-k_3.rs
impl Solution {
    pub fn max_sum_submatrix(matrix: Vec<Vec<i32>>, k: i32) -> i32 {
        
        let m = matrix.len();
        if m == 0 {
            return 0;
        }
        let n = matrix[0].len();
        if n == 0 {
            return 0;
        }
        let mut ans = std::i32::MIN;
        for i in 0..m {
            let mut sum = vec![0; n];
            for j in i..m {
                for c in 0..n {
                    sum[c] += matrix[j][c];
                }
                ans = ans.max(Self::max_sum_subarray(&sum, k));
            }
        }
        ans
    }
    fn max_sum_subarray(nums: &[i32], k: i32) -> i32 {
        let mut ans = std::i32::MIN;
        let mut sum = 0;
        let mut set = std::collections::BTreeSet::new();
        set.insert(0);
        for num in nums {
            sum += num;
            if let Some(&x) = set.range(sum - k..).next() {
                ans = ans.max(sum - x);
            }
            set.insert(sum);
        }
        ans
    }
}